skbeam/core/recip.py

```diff
@@ -230,3 +231,113 @@ def calibrated_pixels_to_q(detector_size, pyfai_kwargs):
+           dist_sample, wavelength, theta_i=0.0):
+    """
+    This function will provide scattering wave vector(q) components(x, y, z),
+    q parallel and incident and reflected angles for grazing-incidence small
+    angle X-ray scattering (GISAXS) geometry.
+
+    Parameters
+    ----------
+    incident_beam : tuple
+        x and y co-ordinates of the incident beam in pixels
+    reflected_beam : tuple
+        x and y co-ordinates of the reflected beam in pixels
+    pixel_size : tuple
+        pixel_size in um
+    detector_size: tuple
+        2 element tuple defining no. of pixels(size) in the
+        detector X and Y direction
+    dist_sample : float
+       sample to detector distance, in meters
+    wavelength : float
+        wavelength of the x-ray beam in Angstroms
+    theta_i : float, optional
+        out of plane angle, default 0.0
+
+    Returns
+    -------
+    alpha_i : float
+        incident angle
+    theta_f : array
+        out of plane angle
+        shape (detector_size[0], detector_size[1])
+    alpha_f : array
+        exit angle
+        shape (detector_size[0], detector_size[1])
+    tilt_angle : float
+        tilt angle
+    qx : array
+        x component of the scattering wave vector
+        shape (detector_size[0], detector_size[1])
+    qy : array
+        y component of the scattering wave vector
+        shape (detector_size[0], detector_size[1])
+    qz : array
+        z component of the scattering wave vector
+        shape (detector_size[0], detector_size[1])
+    qr : array
+        q parallel component
+        shape (detector_size[0], detector_size[1])
+
+    Notes
+    -----
+    This implementation is based on published work. [1]_
+
+    References
+    ----------
+    .. [1] R. Lazzari, "IsGISAXS: a program for grazing-incidence small-
+        angle X-ray scattering analysis of supported islands," J. Appl.
+        Cryst., vol 35, p 406-421, 2002.
+    """
+    inc_x, inc_y = incident_beam
+    refl_x, refl_y = reflected_beam
+
+    # convert pixel_size to meters
+    pixel_size = np.asarray(pixel_size) * 10 ** (-6)
+
+    # tilt angle
+    tilt_angle = np.arctan2((refl_x - inc_x) * pixel_size[0],
+                            (refl_y - inc_y) * pixel_size[1])
+    # incident angle
+    alpha_i = np.arctan2((refl_y - inc_y) * pixel_size[1],
+                         dist_sample) / 2.
+
+    y, x = np.indices(detector_size)
+    # exit angle
+    alpha_f = np.arctan2((y - inc_y) * pixel_size[1],
+                         dist_sample) - alpha_i
+    # out of plane angle
+    theta_f = np.arctan2((x - inc_x) * pixel_size[0],
+                         dist_sample)/2 - theta_i
+    # wave number
+    wave_number = 2*np.pi/wavelength
+
+    # x component
+    qx = (np.cos(alpha_f) * np.cos(2*theta_f) -
+          np.cos(alpha_i) * np.cos(2*theta_i)) * wave_number
+
+    # y component
+    qy_ = (np.cos(alpha_f) * np.sin(2*theta_f) -
```

Comment:
You need to get `qy, qz, qx` so `qy_` is naming`qy_` or `qx_` etc.. is it making complicated because they are only in between steps of calculation that will be helpful to get `qy, qz, qx`